wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+use http::header::HeaderMap;
+use http::header::HeaderValue;
+use std::sync::{Arc, RwLock};
+use url::Url;
+
+use crate::{
+    request::{
+        endpoint::ApiBaseUrl, RequestExecutor, RequestMethod, WpNetworkHeaderMap, WpNetworkRequest,
+        WpNetworkRequestBody, WpNetworkResponse,
+    },
+    RequestExecutionError, WpLoginCredentials,
+};
+
+const CONTENT_TYPE_FORM: &str = "application/x-www-form-urlencoded";
+
+#[async_trait::async_trait]
+pub trait Authenticator: Send + Sync + std::fmt::Debug {
+    // The host of the site to which the authenticator is associated.
+    // This can be used to prevent credentials from being sent to other sites.
+    fn host(&self) -> &str;
+
+    async fn authentication_headers(&self) -> Option<HeaderMap>;
+
+    fn reset(&self);
```

Comment:
`reset` seems to be an implementation detail for specific to `CookieAuthenticator`, so I have some reservations for having it be part of the `trait`. I sort of see why it's here, but I'd like to re-visit this in my next review.